src/lib.rs

```diff
@@ -0,0 +1,1036 @@
+/*
+ * The following license applies to this file, which derives many
+ * details (register and constraint definitions, for example) from the
+ * files `BacktrackingAllocator.h`, `BacktrackingAllocator.cpp`,
+ * `LIR.h`, and possibly definitions in other related files in
+ * `js/src/jit/`:
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#![allow(dead_code)]
+
+pub mod bitvec;
+pub(crate) mod cfg;
+pub(crate) mod domtree;
+pub(crate) mod ion;
+pub(crate) mod moves;
+pub(crate) mod postorder;
+pub(crate) mod ssa;
+
+#[macro_use]
+mod index;
+pub use index::{Block, Inst, InstRange, InstRangeIter};
+
+pub mod checker;
+
+#[cfg(feature = "fuzzing")]
+pub mod fuzzing;
+
+/// Register classes.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum RegClass {
+    Int = 0,
+    Float = 1,
+}
+
+/// A physical register. Contains a physical register number and a class.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct PReg(u8, RegClass);
+
+impl PReg {
+    pub const MAX_BITS: usize = 5;
+    pub const MAX: usize = (1 << Self::MAX_BITS) - 1;
+    pub const MAX_INDEX: usize = 1 << (Self::MAX_BITS + 1); // including RegClass bit
+
+    /// Create a new PReg. The `hw_enc` range is 6 bits.
+    #[inline(always)]
+    pub fn new(hw_enc: usize, class: RegClass) -> Self {
```

Comment:
It would be nice to be able to make this a `const fn` so that we can define constants like this:
```rust
const ZERO_REG: PReg = PReg::new(31, RegClass::Int);
```

However this requires removing the assert since panics in const fn are still unstable. Perhaps it could be replaced with a `debug_assert` in `hw_enc`?